### PR TITLE
Add measurement docs to metric API

### PR DESCRIPTION
**Pull request by MrAlias**
Based on user feedback, how measurements are made with observable instruments is not immediately clear to new users. This adds documentation intended to help guide these users.

cc @seh

### CODE REVIEW

File: metric/doc.go [hunk 1]
```diff
@@ -68,6 +68,37 @@ asynchronous measurement, a Gauge ([Int64ObservableGauge],
...
+Measurements for synchronous instruments ([Int64Counter], [Int64UpDownCounter],
+[Int64Histogram], [Float64Counter], [Float64UpDownCounter], [Float64Histogram])
+are made using the instrument methods directly. All counter instruments have an
+`Add` method that is used to measure an increment value, and all histogram
+instruments have a `Record` method to measure a data point.
+
+Asynchronous instruments ([Int64ObservableCounter],
+[Int64ObservableUpDownCounter], [Int64ObservableGauge],
+[Float64ObservableCounter], [Float64ObservableUpDownCounter],
+[Float64ObservableGauge]) make measurements within a callback function. The
```
Comment (seh) [inline metric/doc.go:85]: For the synchronous instruments, we said passively that "measurements are made." Do the instruments "make" these measurements, or do they allow us callers to make measurements using the instruments?
```suggestion
and [Float64ObservableGauge]) accept measurements within a callback function. The
```

File: metric/doc.go [hunk 2]
```diff
@@ -68,6 +68,37 @@ asynchronous measurement, a Gauge ([Int64ObservableGauge],
...
+callback is registered with the Meter which ensures the callback is called once
+per collection cycle. A callback can be registered two ways: during its
+creation using an option; the RegisterCallback method of the Meter that created
+the instrument.
```
Comment (seh) [inline metric/doc.go:89]: Is it possible—even if inadvisable—to use both of these techniques together, both registering a callback at instrument creation time _and_ registering one or more callbacks on the `Meter`?
```suggestion
per collection cycle. A callback can be registered in either of two ways: during the instrument's creation using an option, or later with the `RegisterCallback` method of the Meter that created the instrument.
```

File: metric/doc.go [hunk 3]
```diff
@@ -68,6 +68,37 @@ asynchronous measurement, a Gauge ([Int64ObservableGauge],
...
+
+If the following criteria are meet, an option (i.e. [WithInt64Callback],
+[WithFloat64Callback]) can be used during the asynchronous instrument's
+creation to register a callback ([Int64Callback], [Float64Callback]):
```
Comment (seh) [inline metric/doc.go:93]:
```suggestion
creation to register a callback (a [Int64Callback] or [Float64Callback], respectively):
```

File: metric/doc.go [hunk 4]
```diff
@@ -68,6 +68,37 @@ asynchronous measurement, a Gauge ([Int64ObservableGauge],
...
+
+  - The measurement process is known when the instrument is created
+  - Only that instrument will make a measurement within the callback
+  - The callback never needs to be unregistered
+
+If not, use the RegisterCallback method of the Meter that created the
```
Comment (seh) [inline metric/doc.go:99]: It's not obvious when we surround method names in backticks—to set them in monospaced type—and when we leave them as bare words.
```suggestion
If these criteria don't hold, use the RegisterCallback method of the Meter that created the
```

File: metric/meter.go [hunk 5]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
-	// cycle.
+	// Int64ObservableGauge returns a new Int64ObservableGauge instrument
+	// identified by name and configured with options. The instrument is used
+	// to asynchronously record instantaneous int64 measurements once per a
+	// measurement collection cycle.
+	//
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithInt64Callback option to register the callback here, or use the
+	// RegisterCallback method of this Meter to register one later. See the
+	// Measurements section of the package documentation for more information.
```
Comment (seh) [inline metric/meter.go:94]: Nice.

File: metric/doc.go (hunk 1 above)
Comment (MrAlias) [reply metric/doc.go:85]: Yeah, maybe we can tweak the wording here.

I don't think is accept is correct though. The instruments themselves don't accept the measurements.

Maybe say they are used to make measurements?

File: metric/doc.go (hunk 2 above)
Comment (MrAlias) [reply metric/doc.go:89]: Sure. If you want to measure different distinct measurements for the instrument where one measurement meets the criteria and the other doesn't (i.e. multi-instrument callback) there is reason to do so.

File: metric/doc.go (hunk 4 above)
Comment (MrAlias) [reply metric/doc.go:99]: > It's not obvious when we surround method names in backticks—to set them in monospaced type—and when we leave them as bare words.

Go docs do not support backticks, we should only use them in markdown.

File: metric/doc.go [hunk 6]
```diff
@@ -68,6 +68,37 @@ asynchronous measurement, a Gauge ([Int64ObservableGauge],
 See the [OpenTelemetry documentation] for more information about instruments
 and their intended use.
 
+# Measurements
+
+Measurements are made by recording values and information about the values with
+an instrument. How these measurements are recorded depends on the instrument.
+
+Measurements for synchronous instruments ([Int64Counter], [Int64UpDownCounter],
+[Int64Histogram], [Float64Counter], [Float64UpDownCounter], [Float64Histogram])
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply metric/doc.go:77]: This will need to be applied elsewhere as well.

File: metric/doc.go (hunk 1 above)
Comment (MrAlias) [reply metric/doc.go:85]: I think "record measurements" is the correct verb

Comment (MrAlias) [reply metric/doc.go]: Updated, PTAL

File: metric/doc.go (hunk 2 above)
Comment (MrAlias) [reply metric/doc.go]: Updated, PTAL.

File: metric/doc.go (hunk 4 above)
Comment (seh) [reply metric/doc.go]: Oh, that. I have a hard time remembering where we settled with golang/go#51082.

File: metric/doc.go [hunk 7]
```diff
@@ -53,21 +53,54 @@ categories to use.
...
 
+# Measurements
+
+Measurements are made by recording values and information about the values with
+an instrument. How these measurements are recorded depends on the instrument.
+
+Measurements for synchronous instruments ([Int64Counter], [Int64UpDownCounter],
+[Int64Histogram], [Float64Counter], [Float64UpDownCounter], and
+[Float64Histogram]) are recorded using the instrument methods directly. All
+counter instruments have an `Add` method that is used to measure an increment
```
Comment (seh) [inline metric/doc.go:79]: Here and on the next line, we surround method names with backticks. Per golang/go#41958, it sounds like the use here isn't a supported style hint, though that's not using balanced backticks; it's using a backtick or grave accent followed by an apostrophe.

Comment (MrAlias) [reply metric/doc.go]: Ah, good catch. Updated.

**Review by seh — APPROVED** [review]

File: metric/meter.go [hunk 8]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
-	// int64 measurements during a computational operation.
+	// Int64UpDownCounter returns a new Int64UpDownCounter instrument
+	// identified by name and configured with options. The instrument is used
+	// to synchronously record int64 measurements during a computational
+	// operation.
 	Int64UpDownCounter(name string, options ...Int64UpDownCounterOption) (Int64UpDownCounter, error)
-	// Int64Histogram returns a new instrument identified by name and
-	// configured with options. The instrument is used to synchronously record
-	// the distribution of int64 measurements during a computational operation.
+	// Int64Histogram returns a new Int64Histogram instrument identified by
```
Comment (pellared) [inline metric/meter.go:61]:
```suggestion
	// Int64Histogram returns a new [Int64Histogram] instrument identified by
```

File: metric/meter.go [hunk 9]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
-	// Int64ObservableCounter returns a new instrument identified by name and
-	// configured with options. The instrument is used to asynchronously record
-	// increasing int64 measurements once per a measurement collection cycle.
+	// Int64ObservableCounter returns a new Int64ObservableCounter identified
+	// by name and configured with options. The instrument is used to
+	// asynchronously record increasing int64 measurements once per a
+	// measurement collection cycle.
+	//
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithInt64Callback option to register the callback here, or use the
```
Comment (pellared) [inline metric/meter.go:72]:
```suggestion
	// the [WithInt64Callback] option to register the callback here, or use the
```

File: metric/meter.go [hunk 10]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
+	// RegisterCallback method of this Meter to register one later. See the
+	// Measurements section of the package documentation for more information.
 	Int64ObservableUpDownCounter(name string, options ...Int64ObservableUpDownCounterOption) (Int64ObservableUpDownCounter, error)
-	// Int64ObservableGauge returns a new instrument identified by name and
-	// configured with options. The instrument is used to asynchronously record
-	// instantaneous int64 measurements once per a measurement collection
-	// cycle.
+	// Int64ObservableGauge returns a new Int64ObservableGauge instrument
```
Comment (pellared) [inline metric/meter.go:86]:
```suggestion
	// Int64ObservableGauge returns a new [Int64ObservableGauge] instrument
```

File: metric/meter.go [hunk 11]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithInt64Callback option to register the callback here, or use the
+	// RegisterCallback method of this Meter to register one later. See the
+	// Measurements section of the package documentation for more information.
 	Int64ObservableGauge(name string, options ...Int64ObservableGaugeOption) (Int64ObservableGauge, error)
 
-	// Float64Counter returns a new instrument identified by name and
-	// configured with options. The instrument is used to synchronously record
-	// increasing float64 measurements during a computational operation.
+	// Float64Counter returns a new Float64Counter instrument identified by
```
Comment (pellared) [inline metric/meter.go:97]:
```suggestion
	// Float64Counter returns a new [Float64Counter] instrument identified by
```

File: metric/meter.go [hunk 12]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
+	Float64UpDownCounter(name string, options ...Float64UpDownCounterOption) (Float64UpDownCounter, error)
+	// Float64Histogram returns a new Float64Histogram instrument identified by
+	// name and configured with options. The instrument is used to
+	// synchronously record the distribution of float64 measurements during a
+	// computational operation.
 	Float64Histogram(name string, options ...Float64HistogramOption) (Float64Histogram, error)
-	// Float64ObservableCounter returns a new instrument identified by name and
-	// configured with options. The instrument is used to asynchronously record
-	// increasing float64 measurements once per a measurement collection cycle.
+	// Float64ObservableCounter returns a new Float64ObservableCounter
```
Comment (pellared) [inline metric/meter.go:112]:
```suggestion
	// Float64ObservableCounter returns a new [Float64ObservableCounter]
```

File: metric/meter.go [hunk 13]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
+	// RegisterCallback method of this Meter to register one later. See the
+	// Measurements section of the package documentation for more information.
+	Float64ObservableUpDownCounter(name string, options ...Float64ObservableUpDownCounterOption) (Float64ObservableUpDownCounter, error)
+	// Float64ObservableGauge returns a new Float64ObservableGauge instrument
+	// identified by name and configured with options. The instrument is used
+	// to asynchronously record instantaneous float64 measurements once per a
+	// measurement collection cycle.
+	//
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithFloat64Callback option to register the callback here, or use the
```
Comment (pellared) [inline metric/meter.go:138]:
```suggestion
	// the [WithFloat64Callback] option to register the callback here, or use the
```

File: metric/meter.go [hunk 14]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
-	// instantaneous int64 measurements once per a measurement collection
-	// cycle.
+	// Int64ObservableGauge returns a new Int64ObservableGauge instrument
+	// identified by name and configured with options. The instrument is used
+	// to asynchronously record instantaneous int64 measurements once per a
+	// measurement collection cycle.
+	//
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithInt64Callback option to register the callback here, or use the
+	// RegisterCallback method of this Meter to register one later. See the
```
Comment (pellared) [inline metric/meter.go:93]:
```suggestion
	// RegisterCallback method of this [Meter] to register one later. See the
```

File: metric/meter.go [hunk 15]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
-	// and configured with options. The instrument is used to asynchronously
-	// record int64 measurements once per a measurement collection cycle.
+	// Int64ObservableUpDownCounter returns a new Int64ObservableUpDownCounter
+	// instrument identified by name and configured with options. The
+	// instrument is used to asynchronously record int64 measurements once per
+	// a measurement collection cycle.
+	//
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithInt64Callback option to register the callback here, or use the
+	// RegisterCallback method of this Meter to register one later. See the
```
Comment (pellared) [inline metric/meter.go:83]:
```suggestion
	// RegisterCallback method of this [Meter] to register one later. See the
```

File: metric/meter.go [hunk 16]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
...
-	// configured with options. The instrument is used to asynchronously record
-	// increasing int64 measurements once per a measurement collection cycle.
+	// Int64ObservableCounter returns a new Int64ObservableCounter identified
+	// by name and configured with options. The instrument is used to
+	// asynchronously record increasing int64 measurements once per a
+	// measurement collection cycle.
+	//
+	// Measurements for the returned instrument are made via a callback. Use
+	// the WithInt64Callback option to register the callback here, or use the
+	// RegisterCallback method of this Meter to register one later. See the
```
Comment (pellared) [inline metric/meter.go:73]:
```suggestion
	// RegisterCallback method of this [Meter] to register one later. See the
```

**Review by pellared — APPROVED** [review]

File: metric/meter.go [hunk 17]
```diff
@@ -49,58 +49,95 @@ type MeterProvider interface {
 type Meter interface {
 	embedded.Meter
 
-	// Int64Counter returns a new instrument identified by name and configured
-	// with options. The instrument is used to synchronously record increasing
-	// int64 measurements during a computational operation.
+	// Int64Counter returns a new Int64Counter instrument identified by name
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply metric/meter.go:52]: These links in an interface method docs won't render.

File: metric/meter.go [hunk 18]
```diff
@@ -55,58 +55,95 @@ type Meter interface {
...
-	// int64 measurements during a computational operation.
+	// Int64UpDownCounter returns a new Int64UpDownCounter instrument
+	// identified by name and configured with options. The instrument is used
+	// to synchronously record int64 measurements during a computational
+	// operation.
 	Int64UpDownCounter(name string, options ...Int64UpDownCounterOption) (Int64UpDownCounter, error)
-	// Int64Histogram returns a new instrument identified by name and
-	// configured with options. The instrument is used to synchronously record
-	// the distribution of int64 measurements during a computational operation.
+	// Int64Histogram returns a new [Int64Histogram] instrument identified by
```
Comment (pellared) [inline metric/meter.go:67]:
```suggestion
	// Int64Histogram returns a new Int64Histogram instrument identified by
```

As it won't render 😬

**Review by pellared — APPROVED** [review]

**Review by hanyuancheung — APPROVED** [review]

**Pull request merged**